### DISCUSSION TITLE
Module 'firedrake' has no attribute 'Cofunction'

### DISCUSSION THREAD

**Discussion by awaham** [firedrakeproject/firedrake#3901] [Firedrake support] [answered]
I am having issues import mesh from firedrake (see below). I am using Ubuntu-20.04 (WSL).

Python 3.12.3 (main, Nov  6 2024, 18:32:19) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from firedrake import mesh
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/firedrake/firedrake/src/firedrake/firedrake/mesh.py", line 37, in <module>
    from firedrake.adjoint_utils import MeshGeometryMixin
  File "/home/firedrake/firedrake/src/firedrake/firedrake/adjoint_utils/__init__.py", line 8, in <module>
    from firedrake.adjoint_utils.function import *               # noqa: F401
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/adjoint_utils/function.py", line 6, in <module>
    from firedrake.adjoint_utils.blocks import FunctionAssignBlock, ProjectBlock, SubfunctionBlock, FunctionMergeBlock, SupermeshProjectBlock
  File "/home/firedrake/firedrake/src/firedrake/firedrake/adjoint_utils/blocks/__init__.py", line 5, in <module>
    from .function import FunctionAssignBlock, FunctionMergeBlock, \
  File "/home/firedrake/firedrake/src/firedrake/firedrake/adjoint_utils/blocks/function.py", line 273, in <module>
    class CofunctionAssignBlock(Block):
  File "/home/firedrake/firedrake/src/firedrake/firedrake/adjoint_utils/blocks/function.py", line 287, in CofunctionAssignBlock
    def __init__(self, lhs: firedrake.Cofunction, rhs: firedrake.Cofunction,
                            ^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'firedrake' has no attribute 'Cofunction'

Any help will be appreciated.

**Comment by connorjward** [comment]
How recent is your Firedrake installation? Can you share the install logs?

**Reply by awaham** [reply]
I did the installation a week ago. Below is the install log.
[firedrake-install.log](https://github.com/user-attachments/files/18002232/firedrake-install.log)

**Reply by connorjward** [reply, marked as answer]
This looks fine. Odd.

Import errors are strange because you should only be dealing with standard Python packages at this point. Can you check:
* That your virtual environment is activated
* That the file `$VIRTUAL_ENV/src/firedrake/firedrake/cofunction.py` exists
* Whether it is possible to run `python -c "from firedrake.cofunction import Cofunction"`.

It might be possible that Python is incorrectly assuming that another folder called `firedrake` (e.g. the venv itself) is a Python package. Sometimes it can be helpful to `cd` to a different directory and try again.

**Reply by jrmaddison** [reply]
I hit this error when I have a `firedrake` subdirectory in the current working directory. Empty `PYTHONPATH` and a fairly standard install on Ubuntu 24.04.

**Reply by dham** [reply]
Argh yes. That's an irritating side effect of editable mode installs.
